package foster.bepb;

message Type {
  enum Tag {
    NAMED = 1;
    PTR = 2;
    PROC = 3;
    CORO = 4;
    TUPLE = 5;
    CARRAY = 6;
    CLOSURE = 7;
    PRIM_INT = 8;
    //FORALL_TY = 9;
    TYPE_VARIABLE = 10;
    ARRAY = 11;
    DATATYPE = 12;
    TY_CON_APP = 13;
  }

  required Tag         tag   = 2;

  // LLVM_NAMED, TYPE_VARIABLE
  optional string name = 3;
  optional ProcType procty = 17;
  repeated Type type_parts = 18;
  optional uint64 carray_size = 19;
  repeated string tyvar_names = 20; // for FORALL_TY
  repeated PbDataCtor ctor    = 21;
}

message PbDataCtor {
  required string   name = 1;
  repeated Type     type = 2;
}

message ProcType {
  repeated Type arg_types = 1;
  required Type ret_type  = 2;
  optional string calling_convention = 3;

  //repeated string annot_key = 3;
  //repeated string annot_val = 4;
}

message Block {
  required string       block_id = 1;
  repeated BlockMiddle  middle   = 2;
  required Terminator   last     = 3;
}

message BlockMiddle {
  optional LetVal      let_val = 1;
  optional LetClosures let_clo = 2;
  optional RebindId    rebind  = 3;
  optional RebindId   bitcast  = 4;
}

message LetVal {
  required string let_val_id = 1;
  required Letable let_expr  = 2;
}

message LetClosures {
  repeated Closure closures = 1;
}

message RebindId {
  required string from_id = 1;
  required TermVar to_var = 2;
}

message Letable {
  enum Tag {
    IL_INT = 16;
    IL_BOOL = 17;
    IL_TUPLE = 20;

    IL_CALL = 23;
    IL_CALL_PRIMOP = 24;
    IL_CTOR  = 25;
    IL_ALLOC = 30;
    IL_DEREF = 31;
    IL_STORE = 32;
    IL_IF = 33;
    IL_ALLOCATE = 34;
    IL_ARRAY_READ = 35;
    IL_ARRAY_POKE = 36;
    IL_NAMED_TYPE_DECL = 37;
  }
  repeated TermVar     parts = 1;
  required Tag           tag = 2;
  optional Type         type = 3;
  repeated string      names = 6;
  optional PBInt      pb_int = 17;

  optional bool   bool_value = 23;   // Bool

  optional PbAllocInfo   alloc_info = 25;
  optional bool call_may_trigger_gc = 26;

  optional PbCoroPrim   coro_prim = 29;
  optional string    prim_op_name = 30;
  optional int32     prim_op_size = 31;
  optional PbCtorId       ctor_id = 32;
}

message Terminator {
  enum Tag {
    BLOCK_RET_VAL  = 1;
    BLOCK_RET_VOID = 2;
    BLOCK_BR       = 3;
    BLOCK_IF       = 4;
    BLOCK_CASE     = 5;
  }
  required Tag         tag = 1;
  optional TermVar var     = 2;  // BLOCK_RET, BLOCK_IF, BLOCK_CASE
  optional string block    = 3;  // BLOCK_IF, BLOCK_BR
  optional string block2   = 4;  // BLOCK_IF
  optional DecisionTree dt = 5;  // BLOCK_CASE
  optional Type        typ = 6;  // BLOCK_CASE
}

message TermVar {
  enum Tag {
    IL_VAR           = 1;
    IL_GLOBAL_SYMBOL = 2;
  }
  required Tag    tag  = 1;
  required string name = 2;
  optional Type   typ  = 3;
}

message PbAllocInfo {
  enum MemRegion {
    MEM_REGION_STACK       = 1;
    MEM_REGION_GLOBAL_HEAP = 2;
  }
  required MemRegion mem_region = 1;
  optional TermVar   array_size = 2;
  required bool      unboxed    = 3;
}

message Proc {
  enum Linkage {
    Internal = 1;
    External = 2;
  }
  required ProcType proctype = 1;
  repeated string   in_args  = 2;
  required string   name     = 3;
  repeated Block    blocks   = 4;
  optional string   lines    = 5;
  required Linkage  linkage  = 6;
}

message Closure {
  required string varname   = 1;
  required string proc_id   = 2;
  required Letable env      = 3;
  required string  env_id   = 4;
}

message PBInt {
  // The middle-end is responsible for verifying that the originalText
  // constitutes a syntactically valid literal. For example, FEEDFACE_10
  // is syntactically invalid.
  //
  // The parser is NOT responsible for verifying that the originalText
  // constitutes a syntactically valid literal, because doing so precludes
  // having __COMPILES__ in the language, which requires transmitting
  // possibly-invalid literals to the middle-end.
  //
  // The middle-end is also responsible for checking that the literal is
  // used appropriately (e.g. no passing 64-bit literal to 32-bit func).
  //
  required string clean = 2; // in base 10
  required int32  bits  = 3;
}

message PbCoroPrim {
  enum Tag {
    IL_CORO_INVOKE = 27;
    IL_CORO_CREATE = 28;
    IL_CORO_YIELD  = 29;
  }
  required Tag       tag = 1;
  required Type ret_type = 2;
  required Type arg_type = 3;
}

message DecisionTree {
  enum Tag {
    DT_FAIL = 2;
    DT_LEAF = 3;
    DT_SWITCH = 4;
  }
  required Tag          tag        = 1;
  optional PbSwitchCase switchcase = 5;
  optional string       leaf_action = 6;
  repeated string       leaf_idents = 7;
  repeated PbOccurrence leaf_idoccs = 8;
}

message PbSwitchCase {
  repeated PbCtorId     ctors   = 1;
  repeated DecisionTree trees   = 2;
  optional DecisionTree defCase = 3;
  optional PbOccurrence occ     = 4;
}

message PbCtorId {
  required string ctor_type_name = 1;
  required string ctor_ctor_name = 2;
  required int32  ctor_local_id  = 3;
}

message PbOccurrence {
  repeated int32    occ_offset = 1;
  repeated PbCtorId occ_ctorid = 2;
}

message Decl {
  required string name = 1;
  required Type   type = 2;
}

message Module {
  required string      modulename = 1;
  repeated Proc        procs      = 2;
  repeated Decl        val_decls  = 3;
  repeated Decl        typ_decls  = 4;
  repeated string      modlines   = 5;
}

